locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsTypeCoercions.java

```diff
@@ -0,0 +1,156 @@
+            for (Map.Entry<?, ?> entry : ((Map<?,?>)input).entrySet()) {
+                String key = (String) entry.getKey();
+                Object val = entry.getValue();
+                
+                Maybe<?> invoked = MethodCoercions.tryFindAndInvokeBestMatchingMethod(builder.get(), key, val);
+                if (invoked == null) {
+                    return Maybe.absent("Builder for "+targetType.getCanonicalName()+" has no matching method "+key);
+                }
+                if (invoked.isAbsent()) {
+                    return (Maybe<T>) (Maybe) invoked;
+                }
+            }
+            
+            Maybe<?> result = MethodCoercions.tryFindAndInvokeMultiParameterMethod(builder.get(), ImmutableList.of(buildMethod.get()), ImmutableList.of());
+            return (Maybe<T>) (Maybe) result;
+        }
+        
+        private Optional<Method> findBuilderMethod(Class<?> clazz) {
+            for (Method m: clazz.getMethods()) {
+                if (((m.getModifiers()&Modifier.STATIC) == Modifier.STATIC) && m.getName().equals("builder")
+                        && m.getParameterTypes().length == 0) {
+                    return Optional.of(m);
+                }
+            }
+            return Optional.empty();
+        }
+        
+        private Optional<Method> findBuildMethod(Object instance) {
+            for (Method m: instance.getClass().getMethods()) {
+                if (((m.getModifiers()&Modifier.STATIC) != Modifier.STATIC) && m.getName().equals("build")
+                        && m.getParameterTypes().length == 0) {
+                    return Optional.of(m);
+                }
+            }
+            return Optional.empty();
+        }
+    }
+    
+    static class CoercionFromAutoValueCreate implements TryCoercer {
+        @Override
+        public <T> Maybe<T> tryCoerce(Object input, TypeToken<T> targetTypeToken) {
+            Class<? super T> targetType = targetTypeToken.getRawType();
+            Maybe<?> result = null;
+            Maybe<?> firstError = null;
+
+            // If we don't have a map of named params, we can't map it to "@SerializedNames"?
+            if (!(input instanceof Map)) {
+                return null;
+            }
+
+            Optional<Method> method = findCreateMethod(targetType);
+            if (!method.isPresent()) return null;
+            
+            // Do we have a map of args, and need to look at the "@SerializedNames" annotation?
+            if (method.get().isAnnotationPresent(SerializedNames.class)) {
+                String[] argNames = method.get().getAnnotation(SerializedNames.class).value();
+                List<Object> orderedArgs = Lists.newArrayList();
+                for (String argName : argNames) {
+                    orderedArgs.add(((Map<?,?>)input).get(argName));
+                }
+                result = MethodCoercions.tryFindAndInvokeMultiParameterMethod(targetType, ImmutableList.of(method.get()), orderedArgs);
```

Comment:
Do you think it's worth failing if there are unmatched inputs, as in the builder case?
Wonder what the better choice is - we skip unmatched templateOptions in general. But if something changes I would prefer an early warning by failing to deploy the blueprint rather than troubleshoot some weird behaviour because the VM was half configured. Also easier to develop if keys are mistyped.